lib/galaxy/tools/parameters/dataset_matcher.py

```diff
@@ -4,8 +4,82 @@
 
 log = getLogger(__name__)
 
-ROLES_UNSET = object()
-INVALID_STATES = [galaxy.model.Dataset.states.ERROR, galaxy.model.Dataset.states.DISCARDED]
+
+def set_dataset_matcher_factory(trans, tool, param_values):
+    trans.dataset_matcher_factory = DatasetMatcherFactory(trans, tool, param_values)
+
+
+def unset_dataset_matcher_factory(trans):
+    trans.dataset_matcher_factory = None
+
+
+def get_dataset_matcher_factory(trans):
+    dataset_matcher_factory = getattr(trans, "dataset_matcher_factory", None)
+    return dataset_matcher_factory or DatasetMatcherFactory(trans)
+
+
+class DatasetMatcherFactory(object):
+    """"""
+
+    def __init__(self, trans, tool=None, param_values=None):
+        self._trans = trans
+        self._tool = tool
+        self._data_inputs = []
+        self._matches_format_cache = {}
+        if tool:
+            valid_input_states = tool.valid_input_states
+        else:
+            valid_input_states = galaxy.model.Dataset.valid_input_states
+        self.valid_input_states = valid_input_states
+        can_process_summary = False
+        if tool is not None and param_values is not None:
+            self._collect_data_inputs(tool, param_values)
+
+            require_public = self._tool and self._tool.tool_type == 'data_destination'
+            if not require_public and self._data_inputs:
+                can_process_summary = True
+                for data_input in self._data_inputs:
+                    if data_input.options:
+                        can_process_summary = False
+                        break
+        self._can_process_summary = can_process_summary
+
+    def matches_any_format(self, hda_extension, formats):
+        for format in formats:
+            if self.matches_format(hda_extension, format):
+                return True
+        return False
+
+    def matches_format(self, hda_extension, format):
+        # cache datatype checking combinations for fast recall
+        if hda_extension not in self._matches_format_cache:
+            self._matches_format_cache[hda_extension] = {}
+
+        formats = self._matches_format_cache[hda_extension]
+        if format not in formats:
+            datatype = self._trans.app.datatypes_registry.get_datatype_by_extension(hda_extension)
+            formats[format] = datatype.matches_any([format])
```

Comment:
`datatype` can be `None`, which would make this fail with `AttributeError: 'NoneType' object has no attribute 'matches_any'` https://travis-ci.org/galaxyproject/galaxy/jobs/371853980#L1995